diesel/src/sqlite/connection/raw.rs

```diff
@@ -292,99 +313,160 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
         //   the memory will have a correct alignment.
         //   (Note I(weiznich): would assume that it is aligned correctly, but we
         //    we cannot guarantee it, so better be safe than sorry)
-        let aggregate_context = ffi::sqlite3_aggregate_context(
-            ctx,
-            std::mem::size_of::<OptionalAggregator<A>>() as i32,
-        );
-        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
-        let aggregator = match aggregate_context.map(|a| &mut *a.as_ptr()) {
+        ffi::sqlite3_aggregate_context(ctx, std::mem::size_of::<OptionalAggregator<A>>() as i32)
+    };
+    let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+    let aggregator = unsafe {
+        match aggregate_context.map(|a| &mut *a.as_ptr()) {
             Some(&mut OptionalAggregator::Some(ref mut agg)) => agg,
             Some(mut a_ptr @ &mut OptionalAggregator::None) => {
                 ptr::write_unaligned(a_ptr as *mut _, OptionalAggregator::Some(A::default()));
                 if let &mut OptionalAggregator::Some(ref mut agg) = a_ptr {
                     agg
                 } else {
-                    unreachable!(
-                        "We've written the aggregator above to that location, it must be there"
-                    )
+                    eprintln!("We've written the aggregator to the aggregate context, but it could not be retrieved");
+                    std::process::abort();
                 }
             }
             None => {
                 null_aggregate_context_error(ctx);
                 return;
             }
-        };
-
-        let mut f = |args: &[*mut ffi::sqlite3_value]| -> Result<(), Error> {
-            let args = build_sql_function_args::<ArgsSqlType, Args>(args)?;
-
-            Ok(aggregator.step(args))
-        };
+        }
+    };
 
-        let args = slice::from_raw_parts(value_ptr, num_args as _);
-        match f(args) {
-            Err(e) => {
-                let msg = e.to_string();
-                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
-            }
-            _ => (),
-        };
-    }
+    let args = unsafe { slice::from_raw_parts(value_ptr, num_args as _) };
+    let args = build_sql_function_args::<ArgsSqlType, Args>(args);
+    let mut aggregator = std::panic::AssertUnwindSafe(aggregator);
```

Comment:
I had some time to think a bit more about this, so here some reasoning why this actually should be OK and users do not need to care about this:

* We catch the panic and get a `Err`
* We return an error message to sqlite, so that sqlite will abort executing this aggregate, so the old instance will not be used anymore.
* On the next call to this aggregate function we will create a new instance of `A`, so we cannot have compromised state here either.

(Maybe add that as comment here)

From that point of view: I'm not sure if we really need to have bounds on `UnwindSafe` on the corresponding types.

Reply:
Ah, I didn't really follow how the aggregator is initialized.
Unwind-safetly is still a useful property though because `finalize()` will be called whether or not `step()` errored out.

I'm not certain if having the `UnwindSafe`-bound actually does much in any case, since this `AssertUnwindSafe` is where the unwind-safety rubber hits the road. Though without it we'd need another `AssertUnwindSafe`-wrapping in `run_aggregator_final_function()` if the bound does not exist. And that may be relevant `Drop` impementations, even if we don't use the aggregator any further.